src/java.base/share/classes/java/lang/reflect/code/op/ExtendedOp.java

```diff
@@ -3155,37 +3155,117 @@ static Block.Builder lowerTypePattern(Block.Builder endNoMatchBlock, Block.Build
                 Block.Builder nextBlock = currentBlock.block();
 
                 // Check if instance of target type
-                Result p;
-                boolean patternWithPrimitive = false;
-                if (targetType instanceof PrimitiveType tt && target.type() instanceof PrimitiveType st) {
-                    patternWithPrimitive = true;
-                    if (SHORT.equals(st) || CHAR.equals(st)) {
-                        st = INT;
-                    }
-                    String s = capitalize(st.toString());
-                    String t = capitalize(tt.toString());
-                    String mn = "is%sTo%sExact".formatted(s, t);
-                    JavaType exactConversionSupport = JavaType.type(ClassDesc.of("java.lang.runtime.ExactConversionsSupport"));
-                    MethodRef mref = MethodRef.method(exactConversionSupport, mn, BOOLEAN, st);
+                Result p; // result of the type check
+                Op c; // op that perform conversion
+                TypeElement s = target.type();
+                TypeElement t = targetType;
+                if (isNarrowingPrimitiveConv(s, t)) {
```

Comment:
I think it would be first simpler to reason about if we characterize whether we are a primitive pattern or not and then further specialize. It should simplify the checking and reduce duplication e.g.,:
```
if (t instanceof PrimitiveType pt) {
  // Primitive type pattern

  PrimitiveType ps;
  if (s instanceof ClassType cs) {
    // Unboxing conversions
  } else {
    // primitive to primitive conversion
    ps = (PrimitiveType) s;
  }

} else {
  // Reference type pattern
}
```
You can still identify specific cases with comments, some i expect should collapse to the same code.

Also we don't need to branch if the predicate would otherwise be a true constant we can just inline.